### DISCUSSION TITLE
Guide for converting exsisting projects?

### DISCUSSION THREAD

**Discussion by bosd** [TomSchimansky/CustomTkinter#188] [Q&A]
There are projects out here which use regular tkinter and could use a fresh gui look.
Is there any guide for converting exsisting projects?

**Comment by ConnyCola** [comment]
Non yet, but I used some regex for trying it out on some of my work projects. Should not be too hard to do that. 
But even nicer would be a compatible import interface. So that you can just import customtkinter as tk and the methods are just all called the same as vanilla tk. All non supported properties could simply be ignored. Any thoughts?

**Reply by bosd** [reply]
Yes, was thinking about regexes too..

Im just a starter in python.
Would it be possible to have it override all tk calls as ctk..

I played around for a couple of hours. With bCNC as a subject. Very cool software, but 1997 gui.
Tried to get to call the update window (with toplevel ctk) and main window with ctk. But without luck.

The best result ive got was that it opened both a tk and ctk window..

At this moment i am at different machine so cant cope the error codes. 
Toplevel complained about some macos window parameter..

**Reply by bosd** [reply]
Toplevel issue in: https://github.com/TomSchimansky/CustomTkinter/issues/190

**Comment by felipetesc** [comment]
We would need to rename all classes to equivalent tkinter class names, all methods would need to have the same method signature from tkinter, and to have several dummy methods with pass keyword. This is what came to my mind, but I'm not quite sure this is all of it, or the only way.

**Reply by felipetesc** [reply]
I almost forgot, the constructor like any other method will need to be changed, and perhaps we would need to add extra methods implementing ctk custom features. Also all ctk features would need default values, obviously

**Reply by bosd** [reply]
Absolutely no expert on this matter, just some thoughts..

- Would it be possible to use an inheritance model?
When ctkinter is loaded, it extends the standard tk methods.
- Any ideas/plans to merge/contribute to regular tkinter?? Their look and feel needs an update to the 2020's. if it is solved in core, it would benefit a lot of projects.

**Reply by felipetesc** [reply]
What I will write here are only my thoughts, this project is Tom's baby, after all, and I'll help him if I can. Sometimes I don't support with more code and new features because I have slightly feeling that I'm steping on his shoes, 'n I don't have this intention.

If it was my decision I would opt to: 
- fix existing bugs for already created widgets
- increase performance
- guarantee security
- enhaencement of the library with new relevant UI elements
- allow ctk to be compatible with tkinter (like that, even if ctk doesn't have a widget, we can use tkinter - it's already possible to mix both )

"Would it be possible to use an inheritance model?
When ctkinter is loaded, it extends the standard tk methods."

If it is possible to use inheritance, yes. It already happens. Ctk, under the hood uses tkinter, and inheritance. For example, there are some important tkinter classes working at the background, they are:
- DrawEngine - no tkinter inheritance here
- CTkCanvas is a chilld of tkinter.Canvas
- CTkBaseClass is a child of tkinter.Frame

"Any ideas/plans to merge/contribute to regular tkinter?? Their look and feel needs an update to the 2020's. if it is solved in core, it would benefit a lot of projects."

I thought about that and it makes sense: "merge/contribute to regular tkinter"
"Their look and feel needs an update to the 2020" - yes, again, I agree with you, it does need an update and this project tries to solve this, I guess
You are correct once more :  "if it is solved in core, it would benefit a lot of projects"

I've never read tkinter source code,  but there are important points about tk: tkinter it is well designed, mature, safe, runs on (Mac/Linux/Windows) and to use as dependency for ctk seems a very smart move from Tom

https://felipetesc.github.io/CtkDocs/#/
